README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# yawp

<!-- badges: start -->

<!-- badges: end -->

A collection of documented, tested, and packaged helper functions.

## Installation

``` r
devtools::install_github("vpnagraj/yawp", branch = "main", build_vignettes = TRUE)
```

Note that the install command above will install the package from GitHub
*and* build vignettes. If you would prefer not to build vignettes upon
install then set `build_vignettes = FALSE`, which is the default in
`devtools::install_github()`.

## Features

  - `betwixt()`: Find values between lower and upper bounds of a range
  - `bound()`: Bound a vector
  - `f1()`: Calculate F1 score
  - `first_char()`: Extract the first character from a string
  - `get_mode()`: Find the mode
  - `gg_zoom()`: Zoom in on a ggplot2 plot (*EXPERIMENTAL*)
  - `mav()`: Calculate moving average
  - `medf()`: Format median
  - `more()`: Add filler rows to a data frame
  - `propf()`: Calculate and format proportion
  - `read_repo()`: Read files directly from a GitHub repository
    (*EXPERIMENTAL*)
  - `summary_se()`: Calculate mean and standard error
  - `theme_mathbook()`: Custom “math book” ggplot2 theme

## Usage

The function documentation includes examples of usage, as do the package
vignettes:

``` r
browseVignettes(package = "yawp")
```

## Inspiration and Related Work

`yawp` is a utility package, and the functionality it provides is by no
means unique. In fact, some of the functions were directly inspired by
other tools. For example:

  - `betwixt()` builds on the concept originally implemented in the
    `between()` function in
    [dplyr](https://CRAN.R-project.org/package=dplyr) but adds a feature
    to ignore specific values
  - `summary_se()` re-implements `summarySE()` from [“The Cookbook for
    R”](http://www.cookbook-r.com/Manipulating_data/Summarizing_data/)
    with tidy eval syntax
  - The moving average functionality in `mav()` is available elsewhere,
    including for example in `slide()` from
    [slider](https://CRAN.R-project.org/package=slider)
  - While `f1()` calculates the F1 score, other available functions such
    as `f_meas()` from
    [yardstick](https://CRAN.R-project.org/package=yardstick) may
    accommodate more generic applications (i.e. applying weights to
    precision and recall)


NEWS.md
# yawp 0.1.2

* Get `names()` of files read in by `read_repo()` to make it possible to return filenames in output tibble.

# yawp 0.1.1

* Added `n_files` argument to `read_repo()`

# yawp 0.1.0

* Initial release!


tests/testthat.R
library(testthat)
library(yawp)

test_check("yawp")


tests/testthat/test-manip.R
test_that("first_char can extract first letter or number", {

  tst <- first_char("NOTHING")
  expect_equal(tst, "N")

  tst <- first_char("compares")
  expect_equal(tst, "c")

  tst <- first_char("2U")
  expect_equal(tst, "2")

})

test_that("logic for finding values between is working", {

  ## by default should return logical indices
  tst <- betwixt(200:300, 250, 255)
  expect_type(tst, "logical")

  ## if value is true should return same type as input
  tst <- betwixt(200:300, 250, 255, value = TRUE)
  expect_type(tst, "integer")
  ## check that returned values are correct (inclusive on both sides)
  expect_equal(tst, c(250,251,252,253,254,255))

  ## check that ignore is working
  tst <- betwixt(200:300, 250, 255, value = TRUE, ignore = 254)
  expect_equal(tst, c(250,251,252,253,255))

  ## check that by is working
  tst <- betwixt(200:300, 250, 255, value = TRUE, by = 2)
  expect_equal(tst, c(250,252,254))

})


tests/testthat/test-stats.R
test_that("correct mode returned", {
  tst <- get_mode(c(1,1,1,1,4,3,3,3,3,3,3,3,3))
  expect_equal(tst, 3)
})

test_that("mode handles ties as expected", {
  ## if ties = TRUE should return both values
  tst <- get_mode(c(1,1,1,1,4,3,2,2,2,2), ties = TRUE)
  expect_equal(tst, c(1,2))

  ## if ties = FALSE should just pick the first of the n tied values
  tst <- get_mode(c(1,1,1,1,4,3,2,2,2,2), ties = FALSE)
  expect_equal(tst, 1)
})

test_that("bound works as expected", {
  ## if ties = TRUE should return both values
  tst <- bound(rnorm(1000, mean = 0.5, sd = 0.5), min = 0.01, max = 0.99)
  expect_true(!any(tst < 0.01))
  expect_true(!any(tst > 0.99))
})

test_that("summary_se returns correct values", {

  ## test expectations based on rounded values from Cookbook for R
  ## "Plotting means and error bars (ggplot2)"
  tst <- summary_se(ToothGrowth, measure_var = len, supp, dose)
  expect_equal(round(tst$mean, 2), c(13.23,22.70,26.06,7.98,16.77,26.14))
  expect_equal(round(tst$se, 2), c(1.41,1.24,0.84,0.87,0.80,1.52))

})

test_that("f1 score is correct for use_thresh = FALSE", {

  ## create same two class data frame used in the yardstick "two_class_example"
  two_class <-
    dplyr::tribble(~ truth, ~ predicted, ~ n,
                   "Class1", "Class1", 227,
                   "Class1", "Class2", 31,
                   "Class2", "Class1", 50,
                   "Class2", "Class2", 192)

  ## calculate f1 score
  tst <-
    data.frame(truth= rep(two_class$truth, two_class$n),
             predicted = rep(two_class$predicted, two_class$n)) %>%
    f1(., observed = predicted, truth = truth, positive = "Class1") %>%
    dplyr::pull(f1)

  ## should equal the documented value from f_meas (with beta = 1, f_meas == f1)
  expect_equal(round(tst,3), 0.849)

})


R/vis.R
#' Zoom in on a ggplot2 plot
#'
#' @description
#'
#' \Sexpr[results=rd, stage=render]{lifecycle::badge("experimental")}]
#'
#' This function takes an input `ggplot2` object and will situate the original side-by-side with the same style plot with a user-defined subset of observations. The effect is that the original plot will be "zoomed" in on specific points.
#'
#' @details
#'
#' **NOTE**: The "zoomed" in plot will be autoscaled based on the values returned, so it may not match the scale of the original. Interpret with caution.
#'
#' @param .plot `ggplot2` plot object that will serve as the original plot to be "zoomed"
#' @param zoom_cmd Expression to return a logical value that, if `TRUE`, will indicate data to keep in the "zoomed" version of the plot; accepts bare column names of the data and is inherited by `...` internally by dplyr \link[dplyr]{filter} to perform the subsetting operation
#' @param draw_box Boolean indicating whether or not to try to draw a rectangle around the subset being zoomed in on for the plot. Default is `TRUE`. Note that this argument will only work if *both* the "x" and "y" axes are on continuous scales.
#' @param box_nudge Coordinates to offset box if `draw_box = TRUE`; nudge applied to min and max of x and y coordinates for the box drawn; default is `1`
#' @param to_label Boolean indicating whether or not the plot should try to label points. Default is `FALSE`.
#' @param label Bare name of the column used to label the data points if `to_label = TRUE`
#'
#' @return A \link[patchwork]{patchwork} object with the `ggplot2` plot and the "zoomed" version side-by-side. The returned object can be further customized using functions from `patchwork` including \link[patchwork]{plot_layout}.
#'
#' @export
#'
#' @md
gg_zoom <- function(.plot, zoom_cmd, draw_box = TRUE, box_nudge = 1, to_label = FALSE, label) {

  ## use enquo for tidyeval syntax
  zoom_cmd <- dplyr::enquo(zoom_cmd)

  ## subset data to zoom in on
  zoom_data <-
    .plot$data %>%
    dplyr::filter(!!zoom_cmd)

  ## build the "zoom plot" based on the original ggplot object
  zoom_plot <- .plot
  ## coerce the data element to be the filtered data
  zoom_plot$data <- zoom_data

  ## if label  arg then add a repel text label
  if(to_label) {

    ## tidyeval syntax allows for a bare column name to be supplied
    label <- dplyr::enquo(label)

    zoom_plot <-
      zoom_plot +
      ggrepel::geom_text_repel(ggplot2::aes(label = !!label))
  }

  ## if draw box then add a box around data used in zoom
  if(draw_box) {

    ## need to get x and y variables (stored as quosures) from ggplot2 object
    x <- .plot$mapping$x
    y <- .plot$mapping$y

    ## create min/max x and y values for box around full plot
    box_data <-
      zoom_data %>%
      dplyr::summarise(
        xmin = min(!!x, na.rm = TRUE),
        xmax = max(!!x, na.rm = TRUE),
        ymin = min(!!y, na.rm = TRUE),
        ymax = max(!!y, na.rm = TRUE)) %>%
      ## add to min and max so that the box sits just outside point
      dplyr::mutate(
        xmin = xmin -  box_nudge,
        xmax = xmax + box_nudge,
        ymin = ymin -  box_nudge,
        ymax = ymax +  box_nudge)

    ## use geom_rect to add to the plot
    .plot <-
      .plot +
      ggplot2::geom_rect(
        ggplot2::aes(xmin = box_data$xmin,
                     xmax = box_data$xmax,
                     ymin = box_data$ymin,
                     ymax = box_data$ymax),
        fill = NA,
        col = "grey",
        lty = "dotted")

  }

  ## return in a basic patchwork layout
  ## patchwork:::`&.gg`(.plot, zoom_plot)
  ## get this function from utils.R
  combine_plots(.plot, zoom_plot)

}


#' Custom "math book" `ggplot2` theme
#'
#' @description
#'
#' This is a custom `ggplot2` theme function that inverts the default white background / black text and customizes legend placement.
#'
#' @details
#'
#' The theme returned from this function uses:
#'
#' - Black panel, plot, and legend backgrounds
#' - White text
#' - Blank legend title
#' - Legend position on bottom
#'
#' @return Returns a \link[ggplot2]{theme} object with custom theme settings for `theme_mathbook`.
#'
#' @export
#'
#' @md
#'
#'
theme_mathbook <- function() {

  ggplot2::theme(plot.background = ggplot2::element_rect(fill = "black",
                                                         color = "black"),
                 panel.background = ggplot2::element_rect(fill = "black",
                                                          color = "black"),
                 legend.background = ggplot2::element_rect(fill = "black"),
                 legend.title = ggplot2::element_blank(),
                 text = ggplot2::element_text(color = "white"),
                 axis.text = ggplot2::element_text(color = "white"),
                 strip.background = ggplot2::element_rect(fill="white"),
                 legend.position = "bottom")

}


R/misc.R
#' Add filler rows to a data frame
#'
#' @description
#'
#' This function is a utility to add "filler" rows to the bottom of a `data.frame`. Basically an ellipsis ("...") for tabular data.
#'
#' @param .data Original `data.frame` or \link[dplyr]{tibble} to which the additional rows will be added; note that if a `data.frame` is passed in it will be coerced to a `tibble` prior to row binding
#' @param fill Character vector of length 1 specifying the content to fill the extra rows; default is `.`
#' @param extra_rows Numeric vector of length 1 indicating how many rows of filler content should be added to the original `.data` object; default is `4`
#'
#' @return A `tibble` with as many rows as the original input (`.data`) plus the number of extra rows (`extra_rows`), all of which are populated with the "filler" (`fill`) content in every column
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' more(head(mtcars))
#' rbind(more(head(mtcars, 1), fill = ".", extra_rows = 1), tail(mtcars, 1))
#'
more <- function(.data, fill = ".", extra_rows = 4) {

  ## force .data to be a tibble
  .data <- tibble::as_tibble(.data)

  ## create a matrix with as many columns as original data ...
  ## ... and as many rows as extra_rows arg
  filler <- matrix(fill, ncol = ncol(.data), nrow = extra_rows)
  ## coerce the matrix to a tibble object
  filler <- tibble::as_tibble(filler, .name_repair = "minimal")
  ## set the column names based on the original data.frame
  names(filler) <- names(.data)

  ## bind together
  rbind(.data, filler)

}

#' Find values between lower and upper bounds of a range
#'
#' @description
#'
#' This is utility function to find values of a vector tha fall between lower and upper bounds specified. The implementation includes a method to optionally ignore certain values and return either a  vector of logical indices or a vector of matching values. Inspired by \link[dplyr]{between}.
#'
#'
#' @param x Vector of numeric or date values
#' @param lower The lower bound (inclusive) for the range of values
#' @param upper The upper bound (inclusive) for the range of values
#' @param ignore Vector containing value(s) to exclude; default is `NULL`
#' @param by Value for increment between lower and upper bounds of the range; default is `1`
#' @param value Boolean as to whether the function should return values of `x` between bounds (`TRUE`) or a vector of logical indices corresponding to whether or not the value falls between bounds (`FALSE`); default is `FALSE`
#'
#' @return
#'
#' Depending on the "value" argument, this will either return a vector of values that fall between the lower and upper bounds *or* a vector of logical vector with `TRUE` for each index corresponding to a value that falls in the range.
#'
#' @export
#'
#' @examples
#'
#' betwixt(1:12, 7, 9)
#' betwixt(1:12, 7, 9, value = TRUE)
#'
#' dates <- seq(as.Date("2000/1/1"), by = "day", length.out = 365)
#' betwixt(dates,
#'         lower = as.Date("2000/1/10"),
#'         upper = as.Date("2000/3/10"),
#'         by = "day",
#'         ignore = as.Date("2000/1/25"),
#'         value = TRUE)
#' @md
betwixt <- function(x, lower, upper, ignore = NULL, by = 1, value = FALSE) {
  ## create sequence from left to right
  ## the by flag allows the user to input increments
  ## NOTE: seq will work on date and numeric objects
  tmp <- seq(from = lower,to = upper,by = by)

  ## if the ignore argument is used then find values of vector
  if(!is.null(ignore)) {
    tmp <- tmp[!tmp %in% ignore]
  }

  ## create vector of boolean indices
  ind <- x %in% tmp
  ## trigger whether to return the boolean TRUE/FALSE ...
  ## ... or the value of the vector at the TRUE indices
  if(value) {
    x[ind]
  } else {
    ind
  }
}


#' Extract the first character from a string
#'
#' @description
#'
#' This helper function will extract the first character from a string. The element may be a letter, number, or special character but will be coerced to a character vector in the output.
#'
#' @param string Character vector from which the first character will be extracted
#'
#' @return Character vector with the first character from each element in the vector passed to the input "string" argument. This value will be the same length as the original vector.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' first_char("purple")
#' first_char(c("purple","rain"))
#' first_char(c("nothing","compares","2u"))
#'
first_char <- function(string) {
  string %>%
    stringr::str_split(., pattern = "") %>%
    purrr::map_chr(., .f = function(x) utils::head(x, 1))
}


#' Bound a vector
#'
#' @description
#'
#' This helper function bounds a numeric vector on a minimum and maximum value.
#'
#' @param x Numeric vector to be bounded
#' @param min Minimum allowed value for vector "x"; default is `0.01`
#' @param max Maximum allowed value for vector "x"; default is `0.99`
#'
#' @return Numeric vector of the same length as `x` with no values less than `minimum` nor greater than `maximum`.
#'
#' @export
#'
#' @md
#' @examples
#'
#' x <- rnorm(1000, 0.5, 0.5)
#' bound(x)
#' bound(x, min = 0.2, max = 0.8)
#' bound(x, min = -0.1, max = 1.1)
#'
bound <- function(x, min = 0.01, max = 0.99) {
  x <- ifelse(x < min, min, x)
  x <- ifelse(x > max, max, x)
  return(x)
}



R/read.R
#' Read files directly from a GitHub repository
#'
#' @description
#'
#' \Sexpr[results=rd, stage=render]{lifecycle::badge("experimental")}
#'
#' This utility allows files from a GitHub repository to be read in bulk. The pattern for file names to read (and sub-directories to traverse) can be arbitrarily defined using a regular expression.
#'
#' @details
#'
#' Starting with a call to the GitHub API, this function returns a list of files, which are then filtered to include only those that match the regular expression in the "pattern" argument. The function then internally constructs the raw content URLs and then applies the read function (".f") to each path.
#'
#' **NOTE**: The API call is unauthenticated. GitHub API request rate limits apply. For specifics, review the GitHub API documentation.
#'
#' @param repo Name of the GitHub repository containing the files
#' @param branch Name of the branch of the GitHub repository containing the files; default is `"master"`
#' @param pattern Pattern to match file names to be returned; accepts \link[base]{regex}
#' @param to_tibble Boolean indicating whether or not the function should attempt to return a `tibble`; default is `FALSE`
#' @param .f Function to read each file whose name complies to `pattern` argument in the repository; default is \link[readr]{read_csv}
#' @param n_files Maximum number of files to read in
#' @param ... Additional arguments passed to the `.f` read function
#'
#' @return Contents of the files in the repository that have read using ".f" function supplied. If `to_tibble = TRUE` then the function will try to stack results on top of each other and return a `tibble`. If `to_tibble = FALSE` then the returned object will be a list with as many elements as there are files in the repository that match the "pattern" argument.
#'
#' @examples
#' \dontrun{
#' read_repo(repo = "cdcepi/Flusight-forecast-data",
#'           branch = "master",
#'           pattern = "data-forecasts/.*/.*\\.csv",
#'           to_tibble = TRUE,
#'           .f = readr::read_csv,
#'           n_files=10,
#'           col_types="DcDccdd",
#'           progress=FALSE,
#'           .id="forecast_filename")
#' }
#' @export
#'
#' @md
#'
read_repo <- function(repo, branch = "master", pattern = NULL, to_tibble = FALSE, .f = readr::read_csv, n_files=NULL, ...) {

  ## construct GET request from the repo and branch
  api_request <- httr::GET(paste0("https://api.github.com/repos/",
                                  repo,
                                  "/git/trees/",
                                  branch,
                                  "?recursive=1"))

  ## extract path element from the API response
  repo_files <- purrr::map_chr(httr::content(api_request)$tree, "path")

  ## construct raw content URLs to files of interest
  repo_files <- file.path("https://raw.githubusercontent.com",
                          repo,
                          branch,
                          repo_files)

  ## if a pattern is passed use it to parse files of interest
  if(!is.null(pattern)) {
    repo_files <- repo_files[grepl(pattern, repo_files)]
  }

  ## Set names
  names(repo_files) <- basename(repo_files)

  ## Limit the number of files
  if (!is.null(n_files) && is.numeric(n_files)) {
    repo_files <- repo_files[1:min(n_files, length(repo_files))]
  }

  ## check that the value passed to .f is a function available in the environment
  .f <- match.fun(.f)

  ## if to_tibble then use map_df() to compile results as a tibble
  ## otherwise return a list via map()
  if(to_tibble) {
    purrr::map_df(repo_files, .f = .f, ...)
  } else {
    purrr::map(repo_files, .f = .f, ...)
  }

}


R/stats.R
#' Calculate mean and standard error
#'
#' @description
#'
#' This function will take an input `data.frame` and summarize a specified numeric column. The summary will include mean, standard error, and a confidence interval corresponding to the width specified (see ".ci" argument). The function leverages tidy eval internally and can accommodate bare column names.
#'
#' @details The inspiration for this function was Winston Chang's `summarySE()` function from "The Cookbook for R". As mentioned in that text, this method of computing standard error is only appropriate for *between* group variation. If there is *within* subject variation (repeated measures) use another method (see references below).
#'
#' @param .data A `data.frame` or `data.frame` extension (e.g. `tibble`)
#' @param measure_var Bare column name for variable to summarize
#' @param ... Optional bare column names for grouping variables
#' @param .ci Width of confidence interval; default is `0.95`
#' @param na.rm Boolean indicating whether or not to use the `na.rm` argument in `mean()` and `sd()` functions; default is `TRUE`
#'
#' @md
#' @return Tibble with columns for grouping variables (if any) and summarized values for mean ("mean"), standard error ("se"), and confidence interval ("ci").
#' @export
#'
#' @references http://www.cookbook-r.com/Manipulating_data/Summarizing_data/
#' @references http://www.cookbook-r.com/Graphs/Plotting_means_and_error_bars_(ggplot2)/#error-bars-for-within-subjects-variables
#' @examples
#' summary_se(.data = ToothGrowth, measure_var = len, supp, dose)
summary_se <- function(.data, measure_var, ..., .ci = 0.95, na.rm = TRUE) {

  ## quote measure variable and grouping variabbles
  measure_var <- dplyr::enquo(measure_var)
  group_var <- dplyr::enquos(...)

  .data %>%
    ## unquote grouping variables (if any)
    dplyr::group_by(!!! group_var) %>%
    ## calculate mean, sd, se, ci
    dplyr::summarise(mean = mean(!! measure_var, na.rm = na.rm),
                     sd = stats::sd(!! measure_var, na.rm = na.rm),
                     n = dplyr::n(),
                     se = sd/sqrt(n),
                     ci = se * stats::qt(.ci/2 + 0.5, n-1),
                     .groups = "drop")

}


#' Format median
#'
#' @description
#'
#' The purpose of this function is to calculate the median, 25th and 75th percentiles of a numeric vector and format output as a string.
#'
#' @details Note that the values for median, 25th, and 75th percentiles are calculated using \link[stats]{fivenum}.
#'
#' @param x Numeric vector to summarize
#' @param digits The number of decimal places to use; default is `2`
#' @param na.rm Boolean indicating whether or not `NA` and `NaN` values should be dropped; inherited by \link[stats]{fivenum}; default is `TRUE`
#' @param ... Additional arguments passed to \link[base]{formatC}
#'
#' @return Character vector of length 1 with formatted median followed by 25th,75th percentiles in parentheses:`"median (25th,75th)"`. The formatting can be further customized using arguments to `base::formatC`.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' x <- sample(1:9, size = 100, replace = TRUE, prob = seq(0.9,0.1,-0.1))
#' medf(x)
#'
medf <- function(x, digits = 2, na.rm = TRUE, ...) {

  ## get 25th, median, 75th percentile
  nums <- stats::fivenum(x, na.rm = na.rm)[2:4]

  paste0(formatC(nums[2], format = "f", digits = digits, ...),
         " (",
         formatC(nums[1], format = "f", digits = digits, ...),
         ",",
         formatC(nums[3], format = "f", digits = digits, ...),
         ")")


}


#' Calculate and format proportion
#'
#' @description
#'
#' This is a simple utility that takes an input vector (character or factor) and calculates proportion of elements in the vector that equal a user-specified value.
#'
#' @details
#'
#' The function works by first converting the input vector "x" to a factor, then the counting the occurrence of the "level" value, calculating the proportion, and finally formatting the output as count and proportion or percentage in parentheses. The default value for "level" is `1`, which assumes the user is interested in calculating the proportion of non-reference values in a binary variable. This utility works just as well with non-binary variables, and the user can issue the function call on the same vector with different values of "level" if there are > 1 non-reference proportions of interest.
#'
#' @param x Vector for which the proportion should of 'level' values should be calculated
#' @param level The non-reference level of interest; default is `1`
#' @param digits The number of decimal places to use; default is `2`
#' @param percent Boolean indicating whether or not the formatted proportion should be converted to percentage; default is `TRUE`
#' @param ... Additional arguments passed to \link[base]{formatC}
#'
#' @return A character vector of length 1 formatted as either `"count (proportion)"` or `"count (percentage%)"` depending on the input to "percent" argument. In either case the formatting can be further customized using arguments to `base::formatC`.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#'
#'dogs <- c("dog","dog","dog")
#'cats <- c("cat","cat")
#'rats <- c("rat","rat","rat","rat","rat")
#'animals <- c(dogs,cats,rats)
#'
#'propf(animals, level = "rat")
#'propf(animals, level = "rat", decimal.mark = ",")
#'propf(animals, level = "dog", percent = FALSE)
#'propf(animals, level = "dog", percent = FALSE, drop0trailing = TRUE)
#'
propf <- function(x, level = "1", digits = 2, percent = TRUE, ...) {

  ## convert input to factor
  x <- as.factor(x)
  ## get the denominator for proportion
  n_total <- length(x)

  ## handle NA values if any
  if (is.na(level) | level == "NA") {
    level <- "NA"
    n_level1 <- length(x[x==level])
  } else {
    n_level1 <- length(x[x==level & !is.na(x)])
  }

  ## calculate proportion
  prop <- n_level1/n_total

  ## format
  if(percent) {
    paste0(n_level1, " (", formatC(100 * prop, format = "f", digits = digits, ...), "%)")
  } else {
    paste0(n_level1, " (", formatC(prop, format = "f", digits = digits, ...), ")")
  }

}

#' Calculate F1 score
#'
#' @description
#'
#' Starting from a `tibble`, this function will calculate binary classification accuracy measures, including the precision and recall, as well as the F1 score, which is the harmonic mean of the two. The input data should include one column for the truth value, and another for the observed. The observed data can either be in the format of predicted outcome or a vector of probabilities. If probabilities are passed to the function, then the user may specify the threshold(s) to classify predicted outcome.
#'
#' @details
#'
#' The following formulas are used to calculate the precision, recall, and F1 score:
#'
#' \deqn{Precision =  TP/(TP+FP)}
#' \deqn{Recall =  TP/(TP+FN)}
#' \deqn{F1 =  2 x ((Precision x Recall) / (Precision + Recall))}
#'
#' Note that F1 is the harmonic mean of precision and recall. The more general F beta score allows precision to be weighted greater than recall or vice versa.
#'
#' @param .data `data.frame` or \link[dplyr]{tibble} with columns for predicted and true values
#' @param observed Bare column name for the predicted value
#' @param truth Bare column name for the true value
#' @param positive Vector of length 1 specifying the value of the 'positive' class in the observed and truth columns; default is `1`
#' @param use_thresh Boolean indicating whether or not the accuracy measures calculated should be based on a threshold(s); this argument should only be set to `TRUE` if the 'observed' column is a vector of probabilities; if `TRUE` the 'thresh' argument will be used to capture thresholds to test; default is `FALSE` and the argument to 'thresh' is ignored
#' @param thresh Vector of thresholds to test; ignored if `use_thresh = FALSE`; default is `0.5`
#'
#' @return A `tibble` with at least one row and four columns: "threshold" (`NA` if `use_thresh = FALSE`), "precision", "recall", "f1". If `use_thresh = TRUE` the `tibble` returned will have as many rows as the length of the vector passed to "thresh".
#'
#' @references Sasaki, Yutaka. (2007). The truth of the F-measure. Teach Tutor Mater.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' fit <- glm(am ~ mpg + wt, data = mtcars, family = "binomial")
#' resp <- predict(fit, newdata = dplyr::select(mtcars, wt, mpg), type = "response")
#' resp <- ifelse(resp > 0.5, 1, 0)
#' dat <- data.frame(am = mtcars$am, prediction = resp)
#'
#' f1(dat, observed = prediction, truth = am, use_thresh = FALSE)
#'
#'
#' x <- rnorm(100, mean = 0.2, sd = 0.4)
#' y <- rnorm(100, mean = 0.6, sd = 0.4)
#' x <- bound(x)
#' y <- bound(y)
#' dat <- data.frame(probs = c(x,y), class = c(rep("x",100), rep("y", 100)))
#'
#' f1(dat, observed = probs, truth = class, use_thresh = TRUE, thresh = c(0.4,0.5,0.6), positive = "x")
#'
f1 <- function(.data, observed, truth, positive = "1", use_thresh = FALSE, thresh = 0.5) {

  observed <- dplyr::enquo(observed)
  truth <- dplyr::enquo(truth)

  ## given how the precision/recall measures  are calculated ...
  ## positive class cannot be '0'
  if(positive == "0") {
    stop("The value passed to the 'positive' argument cannot be '0' ...")
  }

  if(use_thresh) {

    ## vectors to store f1, precision, recall results from loop
    f1_scores <- vector()
    precision_scores <- vector()
    recall_scores <- vector()

    ## loop over each threshold
    ## calculate precision, recall, f1
    for(i in 1:length(thresh)) {

      tmp <-
        .data %>%
        dplyr::filter(!is.na(!!observed)) %>%
        dplyr::mutate(observed = ifelse(!!observed > thresh[i] , positive, "0")) %>%
        # dplyr::count(observed, !!truth) %>%
        dplyr::rename(truth = !!truth)

      ## get n true positives, false positives, false negatives
      n_tp <- nrow(dplyr::filter(tmp, observed == positive & truth == positive))
      n_fp <- nrow(dplyr::filter(tmp, observed == positive & truth != positive))
      n_fn <- nrow(dplyr::filter(tmp, observed != positive & truth == positive))

      ## precision = n true positives divided by (n false positives PLUS  n true positives)
      precision <- n_tp / (n_tp + n_fp)
      ## recall = n true positives divided by (n true positives PLUS n false negatives)
      recall <- n_tp / (n_tp + n_fn)
      ## F1 score 2 times ((precision times recall ) divided by (precision PLUS recall))
      f1 <- 2 * ((precision*recall)/(precision + recall))

      if(length(f1) == 0) f1 <- NA
      if(length(precision) == 0) precision <- NA
      if(length(recall) == 0) recall <- NA

      f1_scores[i] <- f1
      precision_scores[i] <- precision
      recall_scores[i] <- recall

    }
  } else {
    tmp <-
      .data %>%
      dplyr::rename(truth = !!truth,
                    observed = !!observed)

    if(!any(tmp$truth == tmp$observed)) {
      stop(sprintf("None of the values for %s (truth) matched %s (observed) ... ",
                   dplyr::quo_name(truth),
                   dplyr::quo_name(observed)))
    }

    ## get n true positives, false positives, false negatives
    n_tp <- nrow(dplyr::filter(tmp, observed == positive & truth == positive))
    n_fp <- nrow(dplyr::filter(tmp, observed == positive & truth != positive))
    n_fn <- nrow(dplyr::filter(tmp, observed != positive & truth == positive))

    ## precision = n true positives divided by (n false positives PLUS  n true positives)
    precision <- n_tp / (n_tp + n_fp)
    ## recall = n true positives divided by (n true positives PLUS n false negatives)
    recall <- n_tp / (n_tp + n_fn)
    ## F1 score 2 times ((precision times recall ) divided by (precision PLUS recall))
    f1 <- 2 * ((precision*recall)/(precision + recall))

    if(length(f1) == 0) f1 <- NA
    if(length(precision) == 0) precision <- NA
    if(length(recall) == 0) recall <- NA

    ## set thresh to NA because use_thresh = FALSE
    thresh <- NA
    f1_scores <- f1
    precision_scores <- precision
    recall_scores <- recall

  }

  dplyr::tibble(threshold = thresh,
                precision = precision_scores,
                recall = recall_scores,
                f1 = f1_scores)


}


#' Calculate moving average
#'
#' @description
#'
#' This function calculates the moving average of a time series or numeric vector. The rolling calculation can be parameterized by the size of the window with the argument `window` and alignment of the window (centered or left-aligned) with the `align` parameter.
#'
#' @details
#'
#' The functionality here is a wrapper for \link[stats]{filter}. The `align` argument corresponds to `sides` in `stats::filter`, with `align = "left"` triggering `sides = 1` and `align = "center"` triggering `sides = 2`. The function uses the default convolution filter from `stats::filter`.
#'
#' @param x Numeric vector or time series
#' @param window Numeric vector of length 1 indicating the size of window for the moving average
#' @param align Character vector specifying the alignment of the window; allowed values include `"left"` or `"center"`; default is `"center"`
#'
#' @return Vector of the same length as `x` with the moving average at each index. Note `window` - 1 of the values will be `NA`, and depending on the `align` parameter these missing values will appear either at the beginning, end, or both sides of the vector.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' x <- rpois(100, lambda = 3)
#' mav(x, window = 7, align = "center")
#' mav(x, window = 7, align = "left")
mav <- function(x, window, align = "center") {

  if(align == "center") {
    res <- stats::filter(x,rep(1/window,window), sides=2)
  } else if (align == "left") {
    res <- stats::filter(x,rep(1/window,window), sides=1)
  }

  ## if the input was a numeric vector ...
  ## then make sure the output returned is numeric (otherwise a ts)
  if(is.numeric(x)) {
    as.numeric(res)
  } else {
    res
  }
}


#' Find the mode
#'
#' @description
#'
#' This utility calculates the mode (i.e. the most frequently observed value) for a vector of values. The function  function includes an option (see "ties" argument) to return more than one value if there are ties when calculating the mode.
#'
#' @details
#'
#' The mode returned is calculated by tabulating the frequency of each element in the input vector. Note that this method allows for non-numeric vectors to be passed to the function. For example, the function would return the "mode" for a character vector as the most frequently observed string.
#'
#' @param x Vector for which the mode should be calculated
#' @param ties Boolean indicating whether or not ties for mode should be returned; if `FALSE` the function will use \link[base]{which.max} to compute the mode and given ties *only* the first index will be returned; default is `TRUE`
#'
#' @return A vector containing the most frequently observed values in the `x` vector. If there are ties in the mode and the "ties" argument is set to `TRUE`, then the length of the returned vector will be greater than 1. If not, then the length of the returned vector will be 1.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' get_mode(c(1,1,4,3,2,2,2,2))
#' get_mode(c("dog","dog","dog","monkey", "rat"))
#' get_mode(c(1,1,1,1,4,3,2,2,2,2), ties = TRUE)
#' get_mode(c(1,1,1,1,4,3,2,2,2,2), ties = FALSE)
#'
get_mode <- function(x, ties = TRUE) {
  tmp <- unique(x)
  tabulated <- tabulate(match(x,tmp))

  if(ties) {
    tmp[which(tabulated == max(tabulated))]
  } else {
    tmp[which.max(tabulated)]
  }
}
